clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -709,9 +709,8 @@ public boolean hasRoomFor(long timestamp, ByteBuffer key, ByteBuffer value) {
         }
 
         // Be conservative and not take compression of the new record into consideration.
-        return numRecords == 0 ?
-                bufferStream.remaining() >= recordSize :
```

Comment:
Maybe it should be at the top after `isFull()` though since we don't need the other computations in that case.

I think it's reasonable to remove this check since we can't make sure it's always correct. We should probably add a comment to the following code making it clear that it's not really an upper bound.

```java
int size = Math.max(this.batchSize, AbstractRecords.sizeInBytesUpperBound(maxUsableMagic, key, value, headers));
```
We could perhaps fix part of it by taking into account the outer record header for a compressed message batch with message format < V2. It may be worth doing this to avoid confusion.

However, that still doesn't take into account the compression header by Gzip (10 bytes), LZ4 (7 bytes) or Snappy (16 bytes). That's probably OK though as hopefully compression will decrease the size of the key and value so that the record fits. If it doesn't (which may well happen), we will re-allocate the underlying buffer, which is not the end of the world.